File: Python/extrator-desligamentos-cartao-mais-infancia/web/pdf_extractor.py
"""Extração de texto nativo e fallback de OCR para PDFs."""

from __future__ import annotations

import logging
from pathlib import Path

logger = logging.getLogger(__name__)

try:
    import pdfplumber
except ImportError:
    pdfplumber = None

try:
    import pytesseract
    from pdf2image import convert_from_path
except ImportError:
    pytesseract = None
    convert_from_path = None


class PDFExtractor:
    def __init__(self, ocr_enabled: bool = True):
        self.ocr_enabled = ocr_enabled

    def extract(self, pdf_path: str | Path) -> str:
        path = Path(pdf_path)
        if not path.is_file():
            raise FileNotFoundError(path)
        text = self._pdfplumber(path)
        if text.strip() or not self.ocr_enabled:
            return text
        return self._ocr(path)

    @staticmethod
    def _pdfplumber(path: Path) -> str:
        if pdfplumber is None:
            return ""
        try:
            with pdfplumber.open(path) as pdf:
                return "\n".join(page.extract_text() or "" for page in pdf.pages)
        except Exception:
            logger.exception("Falha na extração nativa de %s", path.name)
            return ""

    @staticmethod
    def _ocr(path: Path) -> str:
        if pytesseract is None or convert_from_path is None:
            return ""
        try:
            pages = convert_from_path(str(path))
            return "\n".join(pytesseract.image_to_string(page, lang="por") for page in pages)
        except Exception:
            logger.exception("Falha no OCR de %s", path.name)
            return ""

File: java/extrator-desligamentos-cartao-mais-infancia/backend/src/main/java/com/cmic/desligamentos/service/PdfExtractor.java
package com.cmic.desligamentos.service;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import org.apache.pdfbox.Loader;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.rendering.PDFRenderer;
import org.apache.pdfbox.text.PDFTextStripper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import net.sourceforge.tess4j.Tesseract;

@Service
public class PdfExtractor {

    private static final Logger LOGGER = LoggerFactory.getLogger(PdfExtractor.class);
    private static final float OCR_DPI = 300f;

    public String extract(Path pdfPath, boolean ocrEnabled) {
        try (PDDocument document = Loader.loadPDF(pdfPath.toFile())) {
            String text = nativeText(document);
            if (!text.isBlank() || !ocrEnabled) {
                return text;
            }
            return ocrText(document);
        } catch (IOException exception) {
            LOGGER.error("Falha na extração nativa de {}", pdfPath.getFileName(), exception);
            return "";
        }
    }

    private static String nativeText(PDDocument document) throws IOException {
        PDFTextStripper stripper = new PDFTextStripper();
        return stripper.getText(document);
    }

    private static String ocrText(PDDocument document) {
        try {
            Tesseract tesseract = new Tesseract();
            tesseract.setLanguage("por");
            String datapath = System.getenv("TESSDATA_PREFIX");
            if (datapath != null && !datapath.isBlank()) {
                tesseract.setDatapath(datapath);
            }
            PDFRenderer renderer = new PDFRenderer(document);
            List<String> pages = new ArrayList<>();
            for (int page = 0; page < document.getNumberOfPages(); page++) {
                BufferedImage image = renderer.renderImageWithDPI(page, OCR_DPI);
                pages.add(tesseract.doOCR(image));
            }
            return String.join("\n", pages);
        } catch (Exception exception) {
            LOGGER.error("Falha no OCR", exception);
            return "";
        }
    }
}
